### PR TITLE
Added Neural HardTanh Layer and unit test

### PR DESCRIPTION
**Pull request by arasuarun**
A HardTanh layer could be a nice addition to the neural nets package. It is viewed as an approximation of Tanh:
- 1 if x > 1.0
- -1.0 if x < -1.0
- x otherwise 

The range can be set by the user. Default is (-1, 1).
Not sure of its exact usage (maybe rnn, lstm), but it's present in other dnn frameworks (like torch). 

(Edit:) Some sources:
- Torch [documentation](https://github.com/torch/nn/blob/master/doc/transfer.md#nn.transfer.dok)
- Dr. Yoshua Bengio's [paper](http://arxiv.org/pdf/1206.5533v2.pdf) where it is mentioned ("Practical Recommendations for Gradient-Based Training of Deep Architectures")
- The previous paper refers to this [paper](http://ronan.collobert.com/pub/matos/2004_links_icml.pdf) for the definition.

### CODE REVIEW

File: src/shogun/neuralnets/NeuralHardTanhLayer.h [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	 * being used with
+	 *
+	 */
+	virtual void compute_activations(SGVector<float64_t> parameters,
+		CDynamicObjectArray* layers);
+
+	virtual const char* get_name() const { return "NeuralHardTanhLayer"; }
+
+protected:
+	/** Parameter used to calculate max_val(min_val*(W*x+b),W*x+b).
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: Can you document both parameters seperately?

File: src/shogun/neuralnets/NeuralHardTanhLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+CNeuralHardTanhLayer::CNeuralHardTanhLayer() : CNeuralLinearLayer()
+{
+	m_min_val=-1.0;
+	m_max_val=1.0;
+}
+
+CNeuralHardTanhLayer::CNeuralHardTanhLayer(int32_t num_neurons):
+CNeuralLinearLayer(num_neurons)
+{
+	m_min_val=-1.0;
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralHardTanhLayer.cpp]: Can you register these parameters using SG_ADD? See the other classes

File: src/shogun/neuralnets/NeuralHardTanhLayer.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * 		max_val &\mbox{if } z > max_val \\ 
+ * 		min_val &\mbox{if } z <  min_val \\
+ * 		z& \mbox{otherwise}  
+ * 	\end{cases}
+ * 	\f]
+ * where z=W*x+b and W is the weight matrix, b is the bias vector, x is the input vector, 
+ * and min_val, max_val are parameters.
+ * Default value of min_val is -1.0 and max_val is 1.0.
+ *
+ * When used as an output layer, a squared error measure is used
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: Is there any reference to cite where this is coming from and where one can find evidence that it is useful?

Comment (arasuarun) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: I edited this PR's description to include some sources.

File: src/shogun/neuralnets/NeuralHardTanhLayer.h [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ * Activations are computed according to:
+ * 	\f[ 
+ * 	\begin{cases} 
+ * 		max_val &\mbox{if } z > max_val \\ 
+ * 		min_val &\mbox{if } z <  min_val \\
+ * 		z& \mbox{otherwise}  
+ * 	\end{cases}
+ * 	\f]
+ * where z=W*x+b and W is the weight matrix, b is the bias vector, x is the input vector,
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: Wx+b can be latex as well, as well as all the other math symbols in here

File: tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	input->compute_activations(x);
+	layer.set_min_val(min_val);
+	layer.set_max_val(max_val);
+	layer.compute_activations(params, layers);
+	SGMatrix<float64_t> A = layer.get_activations();
+
+	// manually compute the layer's activations
+	SGMatrix<float64_t> A_ref(layer.get_num_neurons(), x.num_cols);
+
+	float64_t* biases = params.vector;
```
Comment (karlnapf) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: No need for this, SGVector has the [] operator

Comment (karlnapf) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: Ah sorry. I see what you are doing here. But could you still use a SGVector for this? No pointer? You can tell SGVector to not delete the data it received in the constructor

**Comment by karlnapf** [comment]
Very nice and polished PR! Fine to merge from my side apart from the minor things I commented.
Somebody else has to check whether the computation itself makes sense though.

File: src/shogun/neuralnets/NeuralHardTanhLayer.h (hunk 1 above)
Comment (arasuarun) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: Done. I also changed the names from min_val and max_val to min_act and max_act. Thought it made more sense.

File: src/shogun/neuralnets/NeuralHardTanhLayer.cpp (hunk 2 above)
Comment (arasuarun) [inline src/shogun/neuralnets/NeuralHardTanhLayer.cpp]: Done. :) 
But I do notice that aside form the base NeuralLayer class, none of the other Neural*Layer classes have SG_ADD or init(). Should it be added for them, too?

File: tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	{
+		for (int32_t j=0; j<A_ref.num_cols; j++)
+		{
+			A_ref(i,j) = biases[i];
+
+			for (int32_t k=0; k<x.num_rows; k++)
+				A_ref(i,j) += weights[i+k*A_ref.num_rows]*x(k,j);
+
+			A_ref(i,j) = A_ref(i,j) >= max_act ? max_act : 
+				CMath::max<float64_t>(A_ref(i,j),min_act);
```
Comment (karlnapf) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: so many loops .... what about some eigen3/linalg dot products and or openmp?

File: src/shogun/neuralnets/NeuralHardTanhLayer.h [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ * Activations are computed according to:
+ * 	\f[ 
+ * 	\begin{cases} 
+ * 		max_act &\mbox{if } z > max_act \\ 
+ * 		min_act &\mbox{if } z <  min_act \\
+ * 		z& \mbox{otherwise}  
+ * 	\end{cases}
+ * 	\f]
+ * where z=W*x+b and W is the weight matrix, b is the bias vector, x is the input vector, 
+ * and min_act, max_act are parameters.
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: The reference should be included in here, not in the PR description. Nobody reads PR descriptions ...

File: src/shogun/neuralnets/NeuralHardTanhLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+			CMath::max<float64_t>(m_activations[i],m_min_act);
+	}
+}
+
+void CNeuralHardTanhLayer::init()
+{
+	m_min_act = -1.0;
+	m_max_act = 1.0;
+
+	SG_ADD(&m_min_act, "min_act",
```
Comment (karlnapf) [inline src/shogun/neuralnets/NeuralHardTanhLayer.cpp]: Nice. Yeah if you find unregistered parameters, do add them. They are what makes Shogun classes serialisable.
Also make sure that they are initialised properly, otherwise the automatically generated serialisation unit tests will fail in travis

File: tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc (hunk 5 above)
Comment (arasuarun) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: Got it. A nice constructor from SGVector was all that it took.

File: src/shogun/neuralnets/NeuralHardTanhLayer.h (hunk 7 above)
Comment (arasuarun) [inline src/shogun/neuralnets/NeuralHardTanhLayer.h]: Oh! I misunderstood your intention. Handled it. :P

**Comment by arasuarun** [comment]
@karlnapf The third commit should've handled all your comments... expect the linalg/openmp one.

@vigsterkr @karlnapf @lisitsyn To be honest, the entire neuralnets package is filled with *for* loops and (afaik) doesn't use the linalg library much, either. I think it can all be vectorized and refactored with openmp/shogun's linalg. What do you guys think?

Thanks. :)

**Comment by vigsterkr** [comment]
@arasuarun the neuralnet part of shogun was written when the linalg package was starting to emerge. no wonder that it doesn't use all of the linalg backend.

File: src/shogun/neuralnets/NeuralHardTanhLayer.cpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+void CNeuralHardTanhLayer::compute_activations(
+	SGVector<float64_t> parameters,
+	CDynamicObjectArray* layers)
+{
+	CNeuralLinearLayer::compute_activations(parameters, layers);
+
+	int32_t len = m_num_neurons*m_batch_size;
+	for (int32_t i=0; i<len; i++)
```
Comment (vigsterkr) [inline src/shogun/neuralnets/NeuralHardTanhLayer.cpp]: parallelize

File: tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+	layer.compute_activations(params, layers);
+	SGMatrix<float64_t> A = layer.get_activations();
+
+	// manually compute the layer's activations
+	SGMatrix<float64_t> A_ref(layer.get_num_neurons(), x.num_cols);
+
+	SGVector<float64_t> biases = params;
+	SGVector<float64_t> weights(params.vector,params.vlen,layer.get_num_neurons());
+
+	for (int32_t i=0; i<A_ref.num_rows; i++)
```
Comment (vigsterkr) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: parallelize

File: tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+	{
+		for (int32_t j=0; j<A_ref.num_cols; j++)
```
Comment (vigsterkr) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: parallelize

File: tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+		{
+			A_ref(i,j) = biases[i];
+
+			for (int32_t k=0; k<x.num_rows; k++)
```
Comment (vigsterkr) [inline tests/unit/neuralnets/NeuralHardTanhLayer_unittest.cc]: certainly simd

**Comment by karlnapf** [comment]
About loops.

I think we can merge this before putting in openmp.
BUT all linear algebra operations *should* be done with linalg. Never compute dot or matrix products by hand.

Openmp would be a next step. I think this should be done globally in the NN package (spend some time thinking about the design), and for sure a seperate PR. Agree @vigsterkr ?

**Comment by karlnapf** [comment]
Would be a super useful piece of work to parallelise/linalg everything in NNs....

**Comment by vigsterkr** [comment]
why to do things tomorrow when you can do it today? :)

**Comment by arasuarun** [comment]
I'll start by fixing the Tanh and HardTanh layers with linalg and openmp. I'll do so by by the end of this weekend after learning Shogun's openmp conventions and its linalg lib.

**Comment by arasuarun** [comment]
@vigsterkr @karlnapf One thing I was hoping you guys could clear for me: how does Shogun get its GPU support? Afaik, OpenMP isn't for GPUs... right? This is not just for the deep learning part, but in general, I guess.

**Comment by vigsterkr** [comment]
@arasuarun noup openmp is not for gpu: https://en.wikipedia.org/wiki/OpenMP

**Comment by arasuarun** [comment]
Oh! ViennaCL. Got it.

**Comment by sanuj** [comment]
Gpu support comes from ViennaCL.

**Comment by karlnapf** [comment]
news?

**Comment by arasuarun** [comment]
@karlnapf I'm working on this and #3038. My second segment (a third of a semester) ends this weekend, so I have a lot of pending academic work. But I have all of next week off so I'll finish refactoring with linalg and openmp by then. I hope that's fine. :)

**Comment by karlnapf** [comment]
Sure totally fine.
